pkg/v1/tarball/write.go

```diff
@@ -213,3 +249,38 @@ func writeTarEntry(tf *tar.Writer, path string, r io.Reader, size int64) error {
 	_, err := io.Copy(tf, r)
 	return err
 }
+
+type WriteOption func(*writeOptions) error
+type writeOptions struct {
+	updates chan<- v1.Update
+}
+
+func WithProgress(updates chan<- v1.Update) WriteOption {
+	return func(o *writeOptions) error {
+		o.updates = updates
+		return nil
+	}
+}
+
+// progressWriter is a writer which will send the download progress
+type progressWriter struct {
+	w              io.Writer
+	updates        chan<- v1.Update
+	size, complete int64
+}
+
+func (pw progressWriter) Write(p []byte) (int, error) {
+	n, err := pw.w.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	atomic.AddInt64(&pw.complete, int64(n))
+
+	pw.updates <- v1.Update{
```

Comment:
I'm interested to see how this affects performance. It shouldn't block this PR but we should definitely look into it before we add this to crane

Reply:
I would imagine it wouldn't change much of anything; downloading is far slower than sending a message on a channel. But that, of course, is theory, that should be tested (as you said).